src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderCursor.cs

```diff
@@ -464,9 +467,69 @@ public LineBatch GetBatch()
                     throw Contracts.ExceptDecode(batch.Exception, "Stream reading encountered exception");
                 }
 
+                private static class MultiLineReader
+                {
+                    // When reading lines that contain quoted fields, the quoted fields can contain
+                    // '\n' so we we'll need to read multiple lines (multilines) to get all the fields
+                    // of a given row.
+                    public static string ReadMultiLine(TextReader sr, StringBuilder sb, bool ignoreHashLine)
+                    {
+                        string line;
+                        line = sr.ReadLine();
+
+                        // if it was an empty line or if we've reached the end of file (i.e. line = null)
+                        if (string.IsNullOrEmpty(line))
+                            return line;
+
+                        // In ML.NET we filter out lines beginning with // and # at the beginning of the file
+                        // Or lines beginning with // elsewhere in the file.
+                        // Thus, we don't care to check if there's a quoted multiline when the line begins with
+                        // these chars.
+                        if (line[0] == '/' && line[1] == '/')
+                            return line;
+                        if (ignoreHashLine && line[0] == '#')
+                            return line;
+
+                        // Get more lines until the number of quote characters is even
+                        // 2 consecutive quotes are considered scaped quotes
+                        long numOfQuotes = GetNumberOfChars(line, '"');
+                        if (numOfQuotes % 2 == 0)
+                            return line;
+
+                        sb.Clear();
+                        sb.Append(line);
+                        while (numOfQuotes % 2 != 0)
+                        {
+                            line = sr.ReadLine();
+
+                            if (line == null) // If we've reached the end of the file
+                                break; // MYTODO: This could happen if we have an invalid open quote which never closes so we reach the end of the file without properly closing the field, should we throw instead in this case?
+
+                            if(line.Length != 0)
+                                sb.Append(" "); // MYTODO: should we use instead a "\n" in here to separate lines?
```

Comment:
@justinormont [recommended ](https://github.com/dotnet/machinelearning/pull/4584#discussion_r360255342) adding an actual new line here instead of a space. But I'm not sure if it would matter much. Also I wouldn't know when to append "\n" instead of "\r\n", or if it would make any difference.

Reply:
Yep. We'll want newlines instead of spaces.

The same as newlines are different to human readers, the ML can just as well make use of this information. We also have no control over how the user featurizes their dataset in the ML pipeline; the newlines could be important to their processing.

We also would want to maintain round trip read/write of the dataset. Which means we'd have to modify the `TextSaver` to support (or verify current support) for newlines.

Reply:
I've done this in a more recent iteration. To continue discussing about this, it'd be better to do it here:
[https://github.com/dotnet/machinelearning/pull/5125#discussion_r425779339](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425779339)

---
In reply to: [425757137](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425757137) [](ancestors = 425757137)